### PR TITLE
Internal refactoring to avoid unneeded netlink parsing

### PR DESCRIPTION
**Pull request by lmilleri**
Internal refactoring to avoid unneeded netlink parsing
The parsing of some netlink structures is performed only for those vdpa objects
that need to be returned to the user.

### CODE REVIEW

**Comment by lmilleri** [comment]
@amorenoz please review

File: pkg/kvdpa/device.go [hunk 1]
```diff
@@ -242,7 +242,7 @@ func extractBusNameAndMgmtDeviceName(fullMgmtDeviceName string) (busName string,
 }
 
 /*
-GetVdpaDevicesByPciAddress returns the VdpaDevice objects for the given pciAddress
+GetVdpaDevicesByPciAddress returns the VdpaDevice devices for the given pciAddress
```
Comment (amorenoz) [inline pkg/kvdpa/device.go]: is't that redundant? and, kind of inaccurate (assiming we can call golang's structs objects)? Do you like "structs" better? or maybe just ommit the word?

File: cmd/kvdpa-cli/kvdpa-cli.go [hunk 2]
```diff
@@ -110,6 +128,11 @@ func main() {
 					},
 				},
 			},
+			{Name: "get-from-pci",
```
Comment (amorenoz) [inline cmd/kvdpa-cli/kvdpa-cli.go]: Why not use `list --mgmtdev`

File: pkg/kvdpa/device.go [hunk 3]
```diff
@@ -254,7 +254,27 @@ func GetVdpaDevicesByPciAddress(pciAddress string) ([]VdpaDevice, error) {
 		return nil, unix.EINVAL
 	}
 
-	return GetVdpaDevicesByMgmtDev(busName, mgmtDeviceName)
+	msgs, err := GetNetlinkOps().RunVdpaNetlinkCmd(VdpaCmdDevGet, syscall.NLM_F_DUMP, nil)
```
Comment (amorenoz) [inline pkg/kvdpa/device.go]: You are not "making a more selective request down to netlink." This is the same netlink command that is used in `ListVdpaDevices`

What's the difference between this and using `GetVdpaDevicesByMgmtDev`? Only that `ListVdpaDevices()` calls `getBusInfo()`?

If so, a more generic way to solving this without duplicating 90% of the code would be to  add an argument to `ListVdpaDevices` with a bus name and mgtdev name to filter (that can be empty) and forward those arguments down to `parseDevLinkVdpaDevList` who would now do filtering before filling up businformation. Would that work for you?

File: cmd/kvdpa-cli/kvdpa-cli.go (hunk 2 above)
Comment (lmilleri) [reply cmd/kvdpa-cli/kvdpa-cli.go]: oh yes, good point

File: pkg/kvdpa/device.go (hunk 3 above)
Comment (lmilleri) [reply pkg/kvdpa/device.go]: Yes, `getBusInfo()` is the one giving trouble. I'll check your suggestion, thanks

Comment (lmilleri) [reply pkg/kvdpa/device.go]: Removed unneeded public functions `GetVdpaDevicesByMgmtDev` and `GetVdpaDevicesByPciAddress`. `ListVdpaDevices` is the good one

File: pkg/kvdpa/device.go [hunk 4]
```diff
@@ -185,44 +185,30 @@ func GetVdpaDevice(name string) (VdpaDevice, error) {
...
-has the given bus and device names.
-*/
-func GetVdpaDevicesByMgmtDev(busName, devName string) ([]VdpaDevice, error) {
-	result := []VdpaDevice{}
-	devices, err := ListVdpaDevices()
+// ListVdpaDevices returns a list of all requested vdpa devices
+// fullMgmtDeviceName (if not empty) will act as a filter
+// Example1: "pci/0000:01:02" will return all the devices having bus=pci and mgmtdev=0000:01:02
+// Example2: "0000:01:02" will return all the devices having mgmtdev=0000:01:02 (no bus selected)
+func ListVdpaDevices(fullMgmtDeviceName string) ([]VdpaDevice, error) {
```
Comment (amorenoz) [inline pkg/kvdpa/device.go]: This introduces unnecessary API breakage. You achieve the same (with same level of code reuse) with:
```go
func listVdpaDevices(busName, devName string) ([]VdpaDevice, error) {} // Private!

func ListVdpaDevices()  ([]VdpaDevice, error) { return listVdpaDevices("", "") }
func GetVdpaDevicesByMgmtDev (busName, devName string) ([]VdpaDevice, error) { return listVdpaDevices(busName, devName) }
```

**Review by amorenoz — COMMENTED** [review]
Regarding your commit msg, your signing with:
```
Signed-off-by: Leonardo Milleri <lmilleri@lmilleri-thinkpadp1gen3.remote.csb>
```

Is that intentional? Maybe you want to configure the email in your local repo.

File: pkg/kvdpa/device.go (hunk 4 above)
Comment (lmilleri) [reply pkg/kvdpa/device.go]: ovn-k is the only one using a govdpa public function, i.e. `GetVdpaDevicesByPciAddress`. I don't think `GetVdpaDevicesByMgmtDev` is used anywhere. I don't remember why I have introduced `GetVdpaDevicesByPciAddress` in the first place though, it seems a duplication of `GetVdpaDevicesByMgmtDev` . Anyway ovn-k is currently using govdpa version `v0.1.4` not the latest, so shouldn't this prevent any backward compatiblity issues?

Comment (amorenoz) [reply pkg/kvdpa/device.go]: Agree, `GetVdpaDevicesByPciAddress` should go, let's create an issue for that, but it's still unnecessary to change `ListVdpaDevices` or `GetVdpaDevicesByPciAddress`. The API breakage would force us to bump a mayor version with no strong reason.

Also, ovn-k is not the only user of govdpa

Comment (lmilleri) [reply pkg/kvdpa/device.go]: - `GetVdpaDevicesByPciAddress(pciAddress string)` is used by ovn-k, so we shouldn't remove it
- `GetVdpaDevicesByMgmtDev(busName, devName string)` needs to stay for compatibility
- `ListVdpaDevices()` needs to stay for compatibility
- `ListVdpaDevices(fullMgmtDeviceName string)` will become private and CLI will use one of the above

When bumping a new release we'll do the cleaning, do you agree?

Comment (amorenoz) [reply pkg/kvdpa/device.go]: Yes, API should remain untouched for this PR

File: pkg/kvdpa/device.go [hunk 5]
```diff
@@ -197,38 +198,32 @@ GetVdpaDevicesByMgmtDev returns the VdpaDevice objects whose MgmtDev
...
 
-	vdpaDevs, err := parseDevLinkVdpaDevList(msgs)
+	vdpaDevs, err := parseDevLinkVdpaDevList(busName, devName, msgs)
 	if err != nil {
 		return nil, err
 	}
 	return vdpaDevs, nil
 }
 
+func listVdpaDevices() ([]VdpaDevice, error) {
```
Comment (amorenoz) [inline pkg/kvdpa/device.go]: Why do we need this extra single-line private function? It seems only used on one place....

File: cmd/kvdpa-cli/kvdpa-cli.go [hunk 6]
```diff
@@ -26,26 +25,14 @@ const deviceTemplate = ` - Name: {{ .Name }}
...
-		} else {
-			return fmt.Errorf("Invalid management device name %s", c.String("mgmtdev"))
-		}
-		devs, err = vdpa.GetVdpaDevicesByMgmtDev(bus, name)
-		if err != nil {
-			return err
-		}
+	var mgmtDev = c.String("mgmtdev")
+	if mgmtDev != "" {
+		devs, err = vdpa.GetVdpaDevicesByPciAddress(mgmtDev)
```
Comment (amorenoz) [inline cmd/kvdpa-cli/kvdpa-cli.go]: Why do you assume it's PCI? I think this should call `listVdpaDevicesWithBusDevName` directly

**Comment by amorenoz** [comment]
Also, the commit title is a bit misleading as there is no public function change in this PR any more.

File: cmd/kvdpa-cli/kvdpa-cli.go (hunk 6 above)
Comment (lmilleri) [reply cmd/kvdpa-cli/kvdpa-cli.go]: Yes, now I realize the poor decision to expose the public function `GetVdpaDevicesByPciAddress`, as agreed will make the cleanup in the next release of the library.
I should probably call `GetVdpaDevicesByMgmtDev(busName, devName string)`, `listVdpaDevicesWithBusDevName` is a package private function.

File: go.mod [hunk 7]
```diff
@@ -6,7 +6,7 @@ require (
 	github.com/stretchr/testify v1.7.0
 	github.com/urfave/cli/v2 v2.2.0
 	github.com/vishvananda/netlink v1.1.0
-	golang.org/x/sys v0.1.0
```
Comment (amorenoz) [inline go.mod]: Is there any reason why you're upgrading `sys` from 0.1.0 to 0.12.0?

File: pkg/util/util.go [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+import (
+	"errors"
+	"strings"
+)
+
+// ExtractBusAndMgmtDevice extracts the busName and deviceName from a full device address (e.g. pci)
+// example 1: pci/65:0000.1   -> "pci", "65:0000.1",  nil
+// example 2: vdpa_sim        -> "",    "vdpa_sim",   nil
+// example 3: pci/65:0000.1/1 -> "",    "",           err
+func ExtractBusAndMgmtDevice(fullMgmtDeviceName string) (busName string, mgmtDeviceName string, err error) {
```
Comment (amorenoz) [inline pkg/util/util.go]: This is specific to kernel vdpa framework so it should be inside `kvdpa`, not at the same level.

File: go.mod (hunk 7 above)
Comment (lmilleri) [reply go.mod]: `do mod tidy` does it automatically, I can revert it if you want

Comment (amorenoz) [reply go.mod]: I think we should just not hide this updates under unrelated PRs.

Comment (lmilleri) [reply go.mod]: :+1:

File: pkg/util/util.go (hunk 8 above)
Comment (lmilleri) [reply pkg/util/util.go]: :+1:

Comment (amorenoz) [reply pkg/util/util.go]: Just curious why we need a module at all. Why not add this function to the kvdpa module?

Comment (lmilleri) [reply pkg/util/util.go]: I thought utility functions can grow over time

Comment (amorenoz) [reply pkg/util/util.go]: Forcing users to import two modules for a single function seems an overkill right now. Besides, even if utility functions grow so much that they don't fit a single file or "util-foo.go" files, this module cannot be used without kvdpa so I think maybe we could keep them in the same module? WDYT?

Comment (lmilleri) [reply pkg/util/util.go]: Fair enough, is it ok to keep a separate file for util in the kvdpa module?

Comment (amorenoz) [reply pkg/util/util.go]: Sure, we can have `util.go` inside kvpda module with this function (and any further utility func that can makes sense)

**Review by amorenoz — APPROVED** [review]

**Comment by amorenoz** [comment]
Thanks @lmilleri.

**Pull request merged**